### ISSUE TITLE
RFC: PSR policy clarity & compliance choring

### ISSUE THREAD

**Issue by petecooper** [textpattern/textpattern#1965]
From https://github.com/textpattern/textpattern/blob/dev/CONTRIBUTING.md:

>The project follows the [PSR-4](https://www.php-fig.org/psr/psr-4/) and [PSR-12](https://www.php-fig.org/psr/psr-12/) standards with PHP 5.3 style namespacing.

I'm getting my head around https://github.com/PHPCSStandards/PHP_CodeSniffer/ and running it against `dev`. It's brought up some trivial changes such as spacing (e.g., `Line indented incorrectly; expected at least 8 spaces, found 7`) and some not-so-trivial (e.g., `Constants must be uppercase; expected 'TXPINTERFACE' but found 'txpinterface'`).

I'd like to chore through the codebase for the more trivial fixes, starting with spacing. Given there will be a slew of trivial formatting changes across a lot of files, is it preferable to:

i) submit a single PR per file / directory for verification, or…
ii) chunk a few together, or…
iii) do it all one shot, or…
iv) something else?

On the subject of PSR compliance, it would be helpful to know if we're still tracking the same PSR specs. There are currently 14 accepted PSRs (see https://www.php-fig.org/psr/) and our CodeSniffer config (https://github.com/textpattern/textpattern/blob/dev/phpcs.xml) hasn't been touched in 7 years.

Pinging @Bloke @bloatware @jools-r for advice.

**Comment by petecooper** [comment]
Example report for `https://github.com/textpattern/textpattern/blob/dev/.github/txp-checksums.php` :

```
$ php84 /opt/phpcodesniffer/3/phpcs.phar --standard=~/repos/textpattern/textpattern/phpcs.xml ~/repos/textpattern/textpattern/.github/txp-checksums.php 

FILE: /Users/pete/repos/textpattern/textpattern/.github/txp-checksums.php
----------------------------------------------------------------------------------------------------
FOUND 13 ERRORS AND 1 WARNING AFFECTING 6 LINES
----------------------------------------------------------------------------------------------------
  1 | WARNING | [ ] A file should declare new symbols (classes, functions, constants, etc.) and
    |         |     cause no other side effects, or it should execute logic with side effects, but
    |         |     should not do both. The first symbol is defined on line 22 and the first side
    |         |     effect is on line 25.
 22 | ERROR   | [ ] Constants must be uppercase; expected 'TXPINTERFACE' but found 'txpinterface'
 34 | ERROR   | [ ] Constants must be uppercase; expected 'TXPATH' but found 'txpath'
 56 | ERROR   | [x] Expected 1 space after FUNCTION keyword; 0 found
 56 | ERROR   | [x] Opening brace must be the last content on the line
 56 | ERROR   | [x] Closing brace must be on a line by itself
 56 | ERROR   | [x] Each PHP statement must be on a line by itself
 57 | ERROR   | [x] Expected 1 space after FUNCTION keyword; 0 found
 57 | ERROR   | [x] Opening brace must be the last content on the line
 57 | ERROR   | [x] Closing brace must be on a line by itself
 57 | ERROR   | [x] Each PHP statement must be on a line by itself
 61 | ERROR   | [x] Opening brace must be the last content on the line
 61 | ERROR   | [x] Closing brace must be on a line by itself
 61 | ERROR   | [x] Each PHP statement must be on a line by itself
----------------------------------------------------------------------------------------------------
PHPCBF CAN FIX THE 11 MARKED SNIFF VIOLATIONS AUTOMATICALLY
----------------------------------------------------------------------------------------------------

Time: 33ms; Memory: 8MB
```

**Comment by Bloke** [comment]
The spacing/formatting things are probably fine and can be done piecemeal or bunched together. We condense 1-line closures so they're easier to scan than with multi-line blocks, but if the [PSR nobheads say they have to always be expanded](https://www.php-fig.org/psr/psr-12/#7-closures) or face linter wrath then so be it.

Where we will run into difficulties is with capitalisation of constants. That MUST be done lock-stock or it'll break everything. Also, as soon as that change is made, all plugins will break or (at least) issue a warning when they try to check `txpinterface`. The zem_plugin compilation template and toolbelt scripts will also need updating at the same time as they use `txpath` at minimum.

The first warning there, about declarations and side-effects, is daft and applies to regular PHP files for best practice. Common coding convention dictates that all constants and whatnot be defined at the top of the file so they're all contained and easy to find or change in future and don't sprinkle 'magic numbers' throughout code where they might be harder to track down and cause unexpetced bugs.

With a script such as this, that checks command line args and bombs out if they're not set, the linter wants us to move the constant declarations below the checks, because if the conditionals are true, the constant has already been declared and thus potentially "taints" the global namespace and could cause side-effects. But in this case, both checks `die()` immediately and return to the command line (not returning to another routine on the call stack) whereby the used memory is trashed and the code unloaded. So the life of this 'tainted' space is exactly 1 line of code. In my mind, that's a suitable tradeoff for maintaining convention but, of course, for inflexible coding standards that's not good enough, and they'd rather forego convention and maintainability of code for purism.

If you want to silence the warning - entirely your call, I won't lose sleep over it if the constant declaration moves - then shift the first line down to below the 2nd conditional check and it should go away.

Hope that helps.

**Comment by Bloke** [comment]
I don't mind which standard we adopt. There are so many, haha. I think we originally shot for PSR-2 (now PSR-12) but if there's one that's more appropriate then we can always aim for that.

**Comment by bloatware** [comment]
I'm pretty sure there is no standard that would comply with 90% of my code, sorry. But txp5 is a chance for changes, even at bwc price.

**Comment by petecooper** [comment]
Thanks @Bloke, thanks @bloatware.

I've given this some thought - we have a few potential 'wins' (very loosely) if we can restrict CodeSniffer to _just_ indents / spacing (i.e., not code changes) for the remainder of the 4.* release series. This would address numerous Scrutinizer spacing warnings, too…which would reduce the 'noise' from that CI process, which gives clarity to whatever else needs to be looked at.

I understand and absolutely agree with the constants capitalisation situation, and if that is to be addressed then 5.* would be the time to do it. That's way too radical for my current pay grade, and reeks of middle manager meddling.

I was purely thinking about getting us a little closer to the standard we state (PSR-12), and indents / spacing would be a sane-ish place to start. This would leave the code untouched (deliberately so) and reduce the noisiness of the CI reports, among other things.

Not even remotely a hill I'll die on, so if it's a dead end then this can be closed - no drama at all.

**Comment by Bloke** [comment]
Absolutely, the spacing and consistency stuff is worth doing. It's easy to miss so tidying it up is a great move.

**Comment by petecooper** [comment]
Thanks @Bloke. I'll tweak our `phpcs.xml` and get to work.

**Comment by petecooper** [comment]
Closing for now, can revisit nearer 5.0.
